### ISSUE TITLE
FileLocator, realpath, and vfsStream

### ISSUE THREAD

**Issue by rask** [Ocramius/ProxyManager#476]
We are attempting to create an abstraction for proxy generation and we are using this library underneath.

We would also like to test this abstraction.

When I attempt to test proxy generation with vfsStream, there is some internal ProxyManager code that uses `realpath`, which in turn is not vfsStream compatible (https://github.com/bovigo/vfsStream/wiki/Known-Issues).

Is the `realpath()` call a strict requirement in `FileLocator::__construct()`, or could it be altered to allow some custom stream setup to be used when writing proxies to the filesystem?

Would it be worse to skip `realpath` if the given path matches `/^[a-z]+:\/\/.+/`?

This should reproduce the error:

```php
// assume all classes are imported from PM namespaces
function genproxies(string $target_dir)
{
    $configuration = new Configuration();
    $configuration->setProxiesTargetDir($target_dir);
    $configuration->setGeneratorStrategy(new FileWriterGeneratorStrategy(
        new FileLocator($configuration->getProxiesTargetDir())
    ));

    $proxyFactory = new LazyLoadingValueHolderFactory($configuration);

    // ... generate proxies at some point
}

$fs = vfsStream::setup();

genproxies($fs->url());
```

result:

    ProxyManager\Exception\InvalidProxyDirectoryException: Provided directory "vfs://root" does not exist

I assume ProxyManager file generation in general would otherwise support vfsStream, but please let me know if this is not the case.

**Comment by Ocramius** [comment]
The `realpath()` is a hard requirement to avoid dealing with any symlinks.

You can safely swap the `FileLocator` with your own implementation.

**Comment by rask** [comment]
That symlink avoidance sounds like a good candidate for `realpath` usage, yes. Will try and override it for tests.

The file generator implementation is using `tempnam()` which adds another layer to this issue (silently defaults to sys temp dir on stream wrapper URIs). Will see if there are any workarounds for it.

Will continue research, and return in case I think something would require your attention. Thanks!

**Comment by Ocramius** [comment]
> The file generator implementation is using `tempnam()`

This is required: `tempnam()` creates a safe writable file location, which can then be moved, allowing atomic write operations.

**Comment by rask** [comment]
I managed to get vfsStream kinda working on our end. Now I just need to check why only the first test that generates proxies works, while other tests run after that do not.

I just created custom implementations for the interfaces for file location, autoloading, and file writing provided by PM, and injected those in the tests, worked well.

For `tempnam()` I opted for checking if `$location` is a stream, then reinject the stream protocol to the created temp name, and made sure my vfsStream root has a subdir called `tmp` available.
